### PR TITLE
Tab input mode

### LINKED ISSUE #3324: Re-enter insert/passthrough mode after temporarily leaving it
Qutebrowser shouldn't change modes unless the user wants to. As an example of when this occurs, when you're in `passthrough` mode, saving an image by *right click* > *save* will return you to `normal` mode instead of `passthrough`.

Maybe it's possible to have each command, that potentially changes the active mode, restore the previous mode before exit. Either by not storing the state anywhere and passing it as arguments (like in functional programming):

	def my_command(qb_state, ...): # qb_state: current state of qutebrowser, incl. current mode
		... # do something
		return qb_state

Or like this:

	def my_command(qb_state, ...):
		prev_mode = qb_state.mode
		... # do something
		qb_state.mode = prev_mode
		return

Either way, since only one command is executed at a time, the call stack would do all the work. 

Commands, that change tabs, would leave the state alone, allowing each tab to have its own state.

### PR DESCRIPTION
**Pull request by mjauvin**
- save input_mode on mode switch instead of tab focus.
- save mode before a prompt/yesno to be able to restore it once the prompt is closed.
- add tests for all tabs.mode_on_change values

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3590)

*edited by @The-Compiler: Fixes #3324*

### CODE REVIEW

**Comment by The-Compiler** [comment]
Hmm, what's the benefit of saving the input mode on every mode switch if we're not saving it to a session anyways? Also, it looks like a test is failing, at least in [some environments](https://travis-ci.org/qutebrowser/qutebrowser/jobs/340789170#L580).

**Comment by mjauvin** [comment]
Another advantage is that when closing a window with "confirm_quit" set, the prompt will change all windows/tabs back to normal mode... with the _prev_mode change in this PR for modeman, the mode before that global message bill get restored on all windows/tab. This would not be possible if the current input mode had not been saved when actually changing the mode.

**Comment by mjauvin** [comment]
Both tests that failed using ENV=py36-pyqt571 && py36-pyqt58 are passing locally.

File: qutebrowser/keyinput/modeman.py [hunk 1]
```diff
@@ -294,6 +300,9 @@ def leave(self, mode, reason=None, maybe=False):
         self.clear_keychain()
         self.mode = usertypes.KeyMode.normal
         self.left.emit(mode, self.mode, self._win_id)
+        if mode in [usertypes.KeyMode.prompt, usertypes.KeyMode.yesno]:
+            self.enter(self._prev_mode,
+                       reason='restore mode before {}'.format(mode.name))
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeman.py]: This has the same bug your earlier PR had initially: It enters modes like the hint mode without having the necessary preparation.

Also, I really don't think it's worth the complexity: You can see for yourself how easy it is to introduce bugs, and I'm almost sure there are more corner cases when saving modes in different places. I also can't really see many cases where this would be useful (if I'm getting a quit confirmation, I wanted to quit anyways, so I don't care what mode I was in before).

File: qutebrowser/keyinput/modeman.py [hunk 2]
```diff
@@ -266,7 +273,6 @@ def enter_mode(self, mode):
                  usertypes.KeyMode.yesno, usertypes.KeyMode.prompt]:
             raise cmdexc.CommandError(
                 "Mode {} can't be entered manually!".format(mode))
-
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeman.py]: Unrelated whitespace change.

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 3]
```diff
@@ -631,6 +631,16 @@ def on_icon_changed(self, tab, icon):
         if config.val.tabs.tabs_are_windows:
             self.window().setWindowIcon(icon)
 
+    @pyqtSlot(usertypes.KeyMode)
+    def on_mode_entered(self, mode):
+        """Save input mode when tabs.mode_on_change = restore."""
+        input_modes = [usertypes.KeyMode.insert, usertypes.KeyMode.passthrough]
+        if (mode in input_modes and
+                config.val.tabs.mode_on_change == 'restore'):
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py]: This fits on one line without the parentheses :wink:

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 4]
```diff
@@ -631,6 +631,16 @@ def on_icon_changed(self, tab, icon):
...
+            tab = self.currentWidget()
+            if tab and tab is not None:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py]: Why check for both? `if tab is not None:` should be enough.

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 5]
```diff
@@ -642,13 +652,15 @@ def on_mode_left(self, mode):
             if widget is None:
                 return
             widget.setFocus()
+        elif config.val.tabs.mode_on_change == 'restore':
+            tab = self.currentWidget()
+            if tab and tab is not None:
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py]: Same as above.

File: qutebrowser/keyinput/modeman.py (hunk 1 above)
Comment (mjauvin) [reply qutebrowser/keyinput/modeman.py]: I see. The main reason for doing this is that tabs that when using tabs.mode_on_change=restore and passthrough mode is entered on ANY tabs, it will leave this mode whenever ANY prompt is shown (like for certificate/downloads, etc.).

What if I only save here for very specific modes and ONLY when tabs.mode_on_change=restore ?

File: qutebrowser/mainwindow/tabbedbrowser.py (hunk 4 above)
Comment (mjauvin) [reply qutebrowser/mainwindow/tabbedbrowser.py]: Well, maybe I should just do "if tab:" since this page says it *can* be 0 as well:

http://doc.qt.io/qt-5/qtabwidget.html#currentWidget

Comment (The-Compiler) [reply qutebrowser/mainwindow/tabbedbrowser.py]: That's sure an interesting documentation :laughing: 

> The tab dialog does its best to make sure that this value is never 0 (but if you try hard enough, it can be).

However, it's a pointer - and a nullpointer in C++ means `None` in Python.

File: qutebrowser/keyinput/modeman.py (hunk 1 above)
Comment (The-Compiler) [reply qutebrowser/keyinput/modeman.py]: Probably okay to do it then.

File: qutebrowser/keyinput/modeman.py [hunk 6]
```diff
@@ -115,6 +115,7 @@ class ModeManager(QObject):
     Attributes:
         mode: The mode we're currently in.
         _win_id: The window ID of this ModeManager
+        _prev_mode: mode before a prompt poped up
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeman.py]: mode -> Mode, poped -> popped

File: qutebrowser/keyinput/modeman.py [hunk 7]
```diff
@@ -247,6 +250,12 @@ def enter(self, mode, reason=None, only_if_normal=False):
                 return
             log.modes.debug("Overriding mode {}.".format(self.mode))
             self.left.emit(self.mode, mode, self._win_id)
+        if (mode in prompt_modes and self.mode in input_modes and
+                config.val.tabs.mode_on_change == 'restore'):
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeman.py]: I've thought about this some more (and found #3324 again) - I think this should be done independently from `tabs.mode_on_change`.

File: qutebrowser/keyinput/modeman.py [hunk 8]
```diff
@@ -247,6 +250,12 @@ def enter(self, mode, reason=None, only_if_normal=False):
...
+            # save previous mode when being prompted
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeman.py]: I don't think this is an useful comment FWIW :wink:

File: qutebrowser/keyinput/modeman.py [hunk 9]
```diff
@@ -234,6 +236,7 @@ def enter(self, mode, reason=None, only_if_normal=False):
         if mode not in self._parsers:
             raise ValueError("No keyparser for mode {}".format(mode))
         prompt_modes = (usertypes.KeyMode.prompt, usertypes.KeyMode.yesno)
+        input_modes = [usertypes.KeyMode.insert, usertypes.KeyMode.passthrough]
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeman.py]: Given that we need this in a lot of places, we should probably have a `PROMPT_MODES` and `INPUT_MODES` constant in `modeman.py` and use that from other places as well.

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 10]
```diff
@@ -642,13 +651,15 @@ def on_mode_left(self, mode):
             if widget is None:
                 return
             widget.setFocus()
+        elif config.val.tabs.mode_on_change == 'restore':
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py]: The `elif` means the input mode won't be set correctly when `if mode in [usertypes.KeyMode.command, usertypes.KeyMode.prompt, usertypes.KeyMode.yesno]:` is true. I assume that wasn't intended?

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 11]
```diff
@@ -657,26 +668,29 @@ def on_current_changed(self, idx):
...
         tab.setFocus()
 
+        prompt_modes = [usertypes.KeyMode.prompt, usertypes.KeyMode.yesno]
         modes_to_leave = [usertypes.KeyMode.hint, usertypes.KeyMode.caret]
-        if mode_on_change != 'persist':
-            modes_to_leave += modes_to_save
+
+        mm_instance = modeman.instance(self._win_id)
+        current_mode = mm_instance.mode
+        log.modes.debug("{}: mode before tab change: {}."
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py]: That log message is weird to read for a human. What about "Mode before tab change: {} (mode_on_change = {})" or so?

File: qutebrowser/mainwindow/tabbedbrowser.py [hunk 12]
```diff
@@ -657,26 +668,29 @@ def on_current_changed(self, idx):
...
             modeman.leave(self._win_id, mode, 'tab changed', maybe=True)
-        if mode_on_change == 'restore':
-            modeman.enter(self._win_id, tab.data.input_mode,
-                          'restore input mode for tab')
+        if mode_on_change == 'restore' and current_mode not in prompt_modes:
+            modeman.enter(self._win_id, tab.data.input_mode, 'restore')
         if self._now_focused is not None:
             objreg.register('last-focused-tab', self._now_focused, update=True,
                             scope='window', window=self._win_id)
+        log.modes.debug("{}: mode after tab change: {}."
```
Comment (The-Compiler) [inline qutebrowser/mainwindow/tabbedbrowser.py]: Same as above.

File: tests/end2end/features/keyinput.feature [hunk 13]
```diff
@@ -162,3 +162,44 @@ Feature: Keyboard input
         And I press the key "c"
         Then the message "foo 3" should be shown
         And the message "foo 3" should be shown
+
+    # :test all tabs.mode_on_change modes
```
Comment (The-Compiler) [inline tests/end2end/features/keyinput.feature]: I don't understand this comment. Why `:test`?

File: tests/end2end/features/keyinput.feature [hunk 14]
```diff
@@ -162,3 +162,44 @@ Feature: Keyboard input
...
+        And "Leaving mode KeyMode.insert (reason: tab changed)" should be logged
+        And "normal: mode before tab change: insert." should be logged
+        And "normal: mode after tab change: normal." should be logged
+
+    Scenario: mode on change persist
+        Given I set tabs.mode_on_change to persist
+        And I clean up open tabs
+        When I open about:blank
+        And I run :enter-mode insert
+        And I open about:blank in a new tab
```
Comment (The-Compiler) [inline tests/end2end/features/keyinput.feature]: Loading a new page in the foreground leaves insert mode, which breaks this test (and might make others flaky). You probably want to open the page in the background tab and switch tabs instead.

File: qutebrowser/keyinput/modeman.py (hunk 7 above)
Comment (mjauvin) [reply qutebrowser/keyinput/modeman.py]: Oh, I never saw that issue... interesting. So you want to undo these in here and create a different PR for issue #3324 ?

File: qutebrowser/keyinput/modeman.py (hunk 9 above)
Comment (mjauvin) [reply qutebrowser/keyinput/modeman.py]: right, totally agree

File: qutebrowser/keyinput/modeman.py (hunk 7 above)
Comment (The-Compiler) [reply qutebrowser/keyinput/modeman.py]: In the future it'd be good to have different PRs for different futures, but since I already reviewed this and all, feel free to continue it as part of this PR.

File: qutebrowser/mainwindow/tabbedbrowser.py (hunk 11 above)
Comment (mjauvin) [reply qutebrowser/mainwindow/tabbedbrowser.py]: this is intended as a sentinel for tests waiting for log entries (since there is no way to tell when logs from a previous test will end-up in the next test's logs because of the asynchronous nature of qb)

File: tests/end2end/features/keyinput.feature (hunk 13 above)
Comment (mjauvin) [reply tests/end2end/features/keyinput.feature]: no idea... probably a typo. :)

File: qutebrowser/mainwindow/tabbedbrowser.py (hunk 11 above)
Comment (The-Compiler) [reply qutebrowser/mainwindow/tabbedbrowser.py]: It's still also in the debug log and can be useful for debugging issues. Why not make it useful as a sentinel *and* for humans?

Comment (mjauvin) [reply qutebrowser/mainwindow/tabbedbrowser.py]: good point!

File: tests/end2end/features/keyinput.feature (hunk 14 above)
Comment (mjauvin) [reply tests/end2end/features/keyinput.feature]: Not sure what you mean here. Do you mean for mode_on_change=normal or persist? Persist mode should not be leaving insert mode.

**Comment by mjauvin** [comment]
What can I do to move this forward?

Thanks.

**Comment by The-Compiler** [comment]
I still want to look into why the tests are breaking, and whether that's some issue with the code or with the test infrastructure. Unfortunately, my time to review non-trivial pull requests is pretty limited right now (and still will be for the next couple of weeks) - and I've also been sick for over a week now, so even smaller stuff piled up a bit. I'll get to it eventually, though! :wink:

**Comment by jgkamat** [comment]
I took a very quick look at the tests (at least the first failure for `test_mode_on_change_normal`) and it seems  like there's a race condition between leaving insert mode for load started and leaving insert mode for tab changed, which is causing the test to be flakey. In the first situation, that dosen't matter too much so you can just wildcard that, but the second test is a bit more subtle. I suspect that the that's the root cause as well, but I'm not sure how to fix it.

**Comment by The-Compiler** [comment]
I finally got some time to look at PRs and merged this with three changes on top:

- https://github.com/qutebrowser/qutebrowser/commit/11fce30ed0e30c0346c943127a49c8bce39a569a stabilizes the tests by loading different URLs (`about:blank` is a bit of a special case unfortunately), and by not loading them in foreground tabs (as a page which finished loading leaves insert mode, see https://github.com/qutebrowser/qutebrowser/pull/3590#discussion_r174035696)
- https://github.com/qutebrowser/qutebrowser/commit/4dddc077534fb201954d4c240d0bbfd1df392816 makes sure that `modeman.enter(KeyMode.normal)` does the right thing - your code did that, but it never was really "defined", as you'd usually do `modeman.leave` in that case.
- https://github.com/qutebrowser/qutebrowser/commit/07cf2f5b60efc7912dad1ed57299c7adec7cc466 always restores modes when a prompt is closed - I think this will actually be useful even for people who don't have `mode_on_change` set to `restore`.

Thanks!
